### PR TITLE
ww-7 Implement a Login Request to /login and Securely Store API Key a…

### PR DESCRIPTION
**Pull request by katasee**
…nd Token

### CODE REVIEW

File: .DS_Store
Comment (user-has-no-name) [inline .DS_Store:1]: It looks like you still have this file tracked by .git. Delete it.

File: WeWatch/Development.xcconfig [hunk 1]
```diff
@@ -0,0 +1,20 @@
+//
```
Comment (user-has-no-name) [inline WeWatch/Development.xcconfig]: This file should also be added to `.gitignore` to prevent sensitive data such as `API_KEY` from being exposed.

File: WeWatch/Environment.swift [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+//  Envoronment.swift
+//  WeWatch
+//
+//  Created by Anton on 29/11/2024.
+//
+
+import Foundation
+
+public enum Environment {
+    enum  Keys {
```
Comment (user-has-no-name) [inline WeWatch/Environment.swift]: Remove blank space between `enum` and `Keys`

File: WeWatch/Environment.swift [hunk 3]
```diff
@@ -0,0 +1,44 @@
+//
+//  Envoronment.swift
+//  WeWatch
+//
+//  Created by Anton on 29/11/2024.
+//
+
+import Foundation
+
+public enum Environment {
```
Comment (user-has-no-name) [inline WeWatch/Environment.swift]: Add blank line after the first line of the enum.

Comment (user-has-no-name) [inline WeWatch/Environment.swift]: It shouldn't be public. Internal will be enough in such case.

File: WeWatch/Environment.swift [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        }
+        return apiKeyString
+    }()
+    
+    static let apiPin: String = {
+        guard let apiPinKeyString = Environment.infoDictionary[Keys.apiKey] as? String else {
+            fatalError("API Key not set in plist")
+        }
+        return apiPinKeyString
+    }()
```
Comment (user-has-no-name) [inline WeWatch/Environment.swift]: Figure out how to refactor this code to make it more generic and remove duplicates. Create a function that takes a `key` argument and returns the value as a `String`. So that you can call function like `Environment.getPlistValue(key: .apiKey)` and it will return apiKey from plist.

File: WeWatch/Environment.swift [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+import Foundation
+
+public enum Environment {
+    enum  Keys {
+        static let apiKey = "API_KEY"
+        static let baseUrl = "BASE_URL"
+        static let apiPin = "API_PIN"
+    }
+    
+    private static let infoDictionary: [String:Any] = {
```
Comment (user-has-no-name) [inline WeWatch/Environment.swift]: Remember to add space between `[String: Any]`

File: WeWatch/KeychainManager.swift [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+//
+
+
+//
+//  KeychainManager.swift
+//  test KeyChain
+//
+//  Created by Anton on 25/11/2024.
+//
+
```
Comment (user-has-no-name) [inline WeWatch/KeychainManager.swift]: Remove redundant file header.

File: WeWatch/KeychainManager.swift [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+import Foundation
+import Security
+
+class KeychainManager {
```
Comment (user-has-no-name) [inline WeWatch/KeychainManager.swift]: Add access modifiers such as `internal` or `private`

File: WeWatch/KeychainManager.swift [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    
+    enum KeychainError: Error {
+        case duplicateItem
```
Comment (user-has-no-name) [inline WeWatch/KeychainManager.swift:8]: Blank line

File: WeWatch/KeychainManager.swift [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    static func store(data: String, key: String) throws {
+        let data = data.data(using: .utf8)!
+        let query: [String: Any] = [
+            kSecClass as String: kSecClassGenericPassword,
+            kSecAttrAccount as String: key,
+            kSecValueData as String: data,
+            kSecAttrAccessible as String: kSecAttrAccessibleWhenUnlocked
+        ]
+        let status = SecItemAdd(query as CFDictionary, nil)
+        if status == errSecDuplicateItem {
```
Comment (user-has-no-name) [inline WeWatch/KeychainManager.swift]: Add blank lines to make this code more readable.

File: WeWatch/KeychainManager.swift [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+            throw KeychainError.itemNotFound
+        } else if status != errSecSuccess {
+            throw KeychainError.unknown(status)
+        }
+        
+        guard let data = item as? Data else {
+            throw KeychainError.unexpectedDataFormat
+        }
+        return data
+    }
```
Comment (user-has-no-name) [inline WeWatch/KeychainManager.swift]: Add blank lines to make this code more readable.

File: WeWatch/KeychainManager.swift [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+            kSecAttrAccount as String: key,
+        ]
+        let status = SecItemDelete(query as CFDictionary)
+        if status == errSecItemNotFound {
+            throw KeychainError.itemNotFound
+        } else if status != errSecSuccess {
+            throw KeychainError.unknown(status) 
+        }
+    }
+}
```
Comment (user-has-no-name) [inline WeWatch/KeychainManager.swift]: Add blank lines to make this code more readable.

File: WeWatch/WeWatch/ContentView.swift [hunk 12]
```diff
@@ -6,15 +6,52 @@
 //
 
 import SwiftUI
-
 struct ContentView: View {
+    
+    @StateObject private var loginVM = LoginViewModel()
```
Comment (user-has-no-name) [inline WeWatch/WeWatch/ContentView.swift]: Do not use names like loginVM, instead use more explicit names such as `loginViewModel`.

File: WeWatch/WeWatch/Info.plist [hunk 13]
```diff
@@ -2,10 +2,11 @@
 <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
 <plist version="1.0">
 <dict>
-	<key>UIAppFonts</key>
-	<array>
-		<string>Poppins-Regular.ttf</string>
-		<string>Poppins-Bold.ttf</string>
-	</array>
```
Comment (user-has-no-name) [inline WeWatch/WeWatch/Info.plist:9]: Why did you remove the fonts? We'll still need them. Make sure they work properly in the application.

File: WeWatch/WeWatch/ContentView.swift [hunk 14]
```diff
@@ -6,15 +6,52 @@
 //
```
Comment (user-has-no-name) [inline WeWatch/WeWatch/ContentView.swift:6]: Format the code in this file. Look at it and make it as beautiful as you can.

File: WeWatch/WeWatch/LoginViewModel.swift [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+//  LoginViewModel.swift
+//  WeWatch
+//
+//  Created by Anton on 30/11/2024.
+//
+
+import Foundation
+
+class LoginViewModel: ObservableObject {
+    var apikey: String = ""
```
Comment (user-has-no-name) [inline WeWatch/WeWatch/LoginViewModel.swift]: Add blank line

File: WeWatch/WeWatch/LoginViewModel.swift [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+    var pin: String = ""
+    @Published var isLoading: Bool = false
```
Comment (user-has-no-name) [inline WeWatch/WeWatch/LoginViewModel.swift]: Separate your regular properties (like constants and variables) from @Published using blank line

File: WeWatch/WeWatch/Preview Content/Development.xcconfig
Comment (user-has-no-name) [inline WeWatch/WeWatch/Preview Content/Development.xcconfig:1]: This file should contain the same content as Development.xcconfig and must be added to gitignore

File: WeWatch/KeychainManager.swift [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+    
+    enum KeychainError: Error {
+        case duplicateItem
+        case unknown(OSStatus)
+        case itemNotFound
+        case unexpectedDataFormat
+    }
+    
+    static func store(data: String, key: String) throws {
+        let data = data.data(using: .utf8)!
```
Comment (user-has-no-name) [inline WeWatch/KeychainManager.swift]: Avoid using force unwrapping

File: WeWatch/WeWatch/Service/Webservice.swift [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+//
+
+import Foundation
+
+enum AuthenticationError: Error {
+    case invalidCredentials
+    case custom (errorMessage: String)
+}
+
+struct LoginRequestBody: Codable {
```
Comment (user-has-no-name) [inline WeWatch/WeWatch/Service/Webservice.swift]: Remember that if this is the body of the request, you will only need an `Encodable` conformance. Otherwise you will need `Decodable`.

File: WeWatch/WeWatch/Service/Webservice.swift [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+
+struct LoginResponse: Codable {
+    let status: String
+    let data: LoginData?
+}
+
+struct LoginData: Codable {
+    let token: String?
+}
+
```
Comment (user-has-no-name) [inline WeWatch/WeWatch/Service/Webservice.swift]: Remember to add blank line after the first line of the declaration.

File: WeWatch/WeWatch/Service/Webservice.swift [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+class Webservice {
+    func login(apikey: String, pin: String, completion: @escaping (Result<String, AuthenticationError>) -> Void) {
```
Comment (user-has-no-name) [inline WeWatch/WeWatch/Service/Webservice.swift]: If you have multiple arguments in function, you can "break" them to make code more readable. Something like:

```
func login(
    apikey: String,
    pin: String,
    completion: @escaping (Result<String, AuthenticationError>
) -> Void) {
```

Comment (user-has-no-name) [inline WeWatch/WeWatch/Service/Webservice.swift]: Figure out how to make this function more universal. So that you can send requests not only to login, but also to other endpoints.

File: WeWatch/WeWatch/Service/Webservice.swift [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+        
+        URLSession.shared.dataTask(with: request) { (data, response, error) in
+            guard let data = data, error == nil else {
+                DispatchQueue.main.async {
+                    completion(.failure(.custom(errorMessage: "No data or error: \(error?.localizedDescription ?? "Unknown")")))
+                }
+                return
+            }
+            
+            guard let loginResponse = try? JSONDecoder().decode(LoginResponse.self, from: data) else {
```
Comment (user-has-no-name) [inline WeWatch/WeWatch/Service/Webservice.swift]: Do not avoid error handling. If decoding failed - it is better to throw a custom error.

File: WeWatch/WeWatch/Info.plist (hunk 13 above)
Comment (katasee) [reply WeWatch/WeWatch/Info.plist:9]: it's work

File: WeWatch/WeWatch/Service/Webservice.swift [hunk 22]
```diff
@@ -0,0 +1,89 @@
...
+
+struct Resource<T: Codable> {
+    let url: URL
+    var method: HttpMethod = .get([])
+}
+
+class Webservice {
+    
+    func call <T: Codable>(_ resource: Resource<T>) async throws -> T {
+        var request = URLRequest(url: resource.url)
```
Comment (user-has-no-name) [inline WeWatch/WeWatch/Service/Webservice.swift]: Let’s agree to always check your code before pushing to avoid repeating the same issues. Code style: While style isn’t our main focus, it’s important to write clean, well-structured code from the start:
• Remove unnecessary blank lines.
• Add blank lines where they are needed (e.g., after the first line of a struct, enum, or class).
• Always include access modifiers.
• Explicitly specify types wherever possible.

File: WeWatch/WeWatch.xcodeproj/project.pbxproj [hunk 23]
```diff
@@ -284,6 +299,18 @@
...
+				SWIFT_UPCOMING_FEATURE_DISABLE_OUTWARD_ACTOR_ISOLATION = YES;
+				SWIFT_UPCOMING_FEATURE_EXISTENTIAL_ANY = YES;
+				SWIFT_UPCOMING_FEATURE_FORWARD_TRAILING_CLOSURES = YES;
+				SWIFT_UPCOMING_FEATURE_GLOBAL_CONCURRENCY = YES;
+				SWIFT_UPCOMING_FEATURE_IMPLICIT_OPEN_EXISTENTIALS = YES;
+				SWIFT_UPCOMING_FEATURE_IMPORT_OBJC_FORWARD_DECLS = YES;
+				SWIFT_UPCOMING_FEATURE_INFER_SENDABLE_FROM_CAPTURES = YES;
+				SWIFT_UPCOMING_FEATURE_INTERNAL_IMPORTS_BY_DEFAULT = YES;
+				SWIFT_UPCOMING_FEATURE_ISOLATED_DEFAULT_VALUES = YES;
+				SWIFT_UPCOMING_FEATURE_REGION_BASED_ISOLATION = YES;
```
Comment (user-has-no-name) [inline WeWatch/WeWatch.xcodeproj/project.pbxproj:313]: Why would you add it?

File: .gitignore [hunk 24]
```diff
@@ -1,2 +1,8 @@
 **/.DS_Store
 *.xcuserstate
```
Comment (user-has-no-name) [inline .gitignore]: Remove

File: .gitignore [hunk 25]
```diff
@@ -1,2 +1,8 @@
...
+.DS_Store
```
Comment (user-has-no-name) [inline .gitignore]: Remove

File: .gitignore [hunk 26]
```diff
@@ -1,2 +1,8 @@
...
+WeWatch/WeWatch.xcodeproj/project.xcworkspace/xcuserdata/*
+WeWatch/WeWatch.xcodeproj/project.xcworkspace/xcuserdata/katasee.xcuserdatad/UserInterfaceState.xcuserstate
```
Comment (user-has-no-name) [inline .gitignore]: Remove

File: .gitignore [hunk 27]
```diff
@@ -1,2 +1,8 @@
...
+**/xcuserdata/
+WeWatch/Development.xcconfig
+WeWatch/WeWatch/Preview Content/Development.xcconfig
```
Comment (user-has-no-name) [inline .gitignore]: **.xcconfig

File: WeWatch/KeychainManager.swift [hunk 28]
```diff
@@ -0,0 +1,66 @@
...
+
+internal class KeychainManager {
+    enum KeychainError: Error {
+        case duplicateItem
+        case unknown(OSStatus)
+        case itemNotFound
+        case unexpectedDataFormat
+    }
+    
+    internal static func store(data: String, key: String) throws {
```
Comment (user-has-no-name) [inline WeWatch/KeychainManager.swift]:
```
 internal static func store(
    data: String,
    key: String
) throws {
```

File: WeWatch/WeWatch/Service/Webservice.swift [hunk 29]
```diff
@@ -0,0 +1,99 @@
...
+    
+    let token: String?
+}
+
+internal enum  HttpMethod {
+    
+    case get([URLQueryItem])
+    case post(Data?)
+    
+    var name: String {
```
Comment (user-has-no-name) [inline WeWatch/WeWatch/Service/Webservice.swift]: Here you can also can add access modifier

Comment (user-has-no-name) [inline WeWatch/WeWatch/Service/Webservice.swift]: You can also add access modifiers to all properties inside of your structures
